### DISCUSSION TITLE
How to skip a specific version?

### DISCUSSION THREAD

**Discussion by nielsbasjes** [renovatebot/renovate#25862] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

The org.apache.maven.plugins:maven-artifact-plugin version 3.5.0 has a bug that breaks releasing my projects. So I want to disallow this specific version in my renovate config.
So no update to 3.5.0 but as soon as there is a 3.5.1 or 3.6.0 I want to get that update.

With this purpose I have added this to my packageRules:

    {
      "description": "Disable broken maven-artifact-plugin version",
      "matchPackageNames": ["org.apache.maven.plugins:maven-artifact-plugin"],
      "allowedVersions": "!/3.5.0/"
    },

Yet after this renovate still provides me with a merge request for that version.

What have I done incorrectly? 
How do I fix this?

### Logs (if relevant)

Full logs: https://developer.mend.io/github/nielsbasjes/codeowners/-/job/8664b0a5-e0ce-49b2-adf4-08bf0e7fa0e6

It loads my config

```
      {
        "description": "Disable broken maven-artifact-plugin version",
        "matchPackageNames": [
          "org.apache.maven.plugins:maven-artifact-plugin"
        ],
        "allowedVersions": "!/3.5.0/"
      },
```

It finds the newer version (which I do not want)

```
          {
            "currentValue": "3.4.1",
            "currentVersion": "3.4.1",
            "datasource": "maven",
            "depName": "org.apache.maven.plugins:maven-artifact-plugin",
            "depType": "build",
            "fileReplacePosition": 1882,
            "fixedVersion": "3.4.1",
            "groupName": "maven-artifact-plugin.version",
            "homepage": "https://maven.apache.org/plugins/",
            "isSingleVersion": true,
            "packageName": "org.apache.maven.plugins:maven-artifact-plugin",
            "registryUrl": "https://repo.maven.apache.org/maven2",
            "registryUrls": [
              "https://repo.maven.apache.org/maven2"
            ],
            "versioning": "maven",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "3.5.0",
                "newValue": "3.5.0",
                "releaseTimestamp": "2023-09-29T05:52:06.000Z",
                "newMajor": 3,
                "newMinor": 5,
                "updateType": "minor",
                "branchName": "renovate/maven_build_plugins"
              }
            ]
          },
```

and then

```
DEBUG: Updating org.apache.maven.plugins:maven-artifact-plugin in pom.xml (branch="renovate/maven_build_plugins")
...
 "upgrades": [
        {
          "datasource": "maven",
          "depName": "org.apache.maven.plugins:maven-artifact-plugin",
          "displayPending": "",
          "fixedVersion": "3.4.1",
          "currentVersion": "3.4.1",
          "currentValue": "3.4.1",
          "newValue": "3.5.0",
          "newVersion": "3.5.0",
          "packageFile": "pom.xml",
          "updateType": "minor",
          "packageName": "org.apache.maven.plugins:maven-artifact-plugin"
        },
```

**Comment by nielsbasjes** [comment]
Second attempt; same unwanted effect
```
      {
        "description": "Disable broken maven-artifact-plugin version",
        "matchPackageNames": [
          "org.apache.maven.plugins:maven-artifact-plugin"
        ],
        "allowedVersions": "!/^3\\.5\\.0$/"
      },
```

**Reply by PhilipAbed** [reply]
@nielsbasjes 
this worked fine for me
```
 "packageRules": [
       {
      "matchPackageNames": ["@commitlint/cli"],
      "allowedVersions": "!/18.4.2/"
    }
  ]
```
not sure why it didn't for your package
try matching depName instead maybe?

**Reply by PhilipAbed** [reply]
i mean the allowedVersions you had on the first try seems ok, but the `matchPackageNames` is not working as expected for you i believe.
i would try 
https://docs.renovatebot.com/configuration-options/#matchdepnames
https://docs.renovatebot.com/configuration-options/#matchdeppatterns

**Reply by nielsbasjes** [reply]
The matchDepPatterns seems like the regex variant of matchDepNames. 
Since I want the exact name the matchDepNames seems like what I need.

I see in the logs

    "depName": "org.apache.maven.plugins:maven-artifact-plugin",

So I tried

    {
      "description": "Disable broken maven-artifact-plugin version",
      "matchDepNames": ["org.apache.maven.plugins:maven-artifact-plugin"],
      "allowedVersions": "!/^3\\.5\\.0$/"
    },

and no luck.

**Reply by nielsbasjes** [reply]
The actual MR (in case this might give you more info)
https://github.com/nielsbasjes/codeowners/pull/35

**Reply by nielsbasjes** [reply]
Looking at the output in the log I see that both these have the exact same value

    "depName": "org.apache.maven.plugins:maven-artifact-plugin",
    "packageName": "org.apache.maven.plugins:maven-artifact-plugin",

Reading through https://github.com/renovatebot/renovate/issues/16012 I read that the depName is (going to be?) intended as `"pretty name"/"short name"`

So shouldn't I use the matchPackageNames instead of the matchDepNames in this case?

**Reply by PhilipAbed** [reply]
true, but i had a case where depName worked while packageName didn't so i thought its worth a try.
 i tried to reproduce, i couldn't
 
 this is my config:
 https://github.com/PhilipAbed/malicious-dependencies/blob/main/renovate.json
 
i ran, it opened a PR with  maven-plugin-artifact 3.5.0 , but when i added the Package-rule, it removed the PR completely.
![image](https://github.com/renovatebot/renovate/assets/42116482/451c46b2-5d51-4c5a-ae43-1a4898d51e92)

**Reply by PhilipAbed** [reply]
![image](https://github.com/renovatebot/renovate/assets/42116482/35fe88fc-e3fd-461d-b6c3-a75f8a0d9648)

**Reply by PhilipAbed** [reply, marked as answer]
could be another rule after it overriding it? try to set it at the end of the config! and do it like i did(see example above in config)

**Reply by nielsbasjes** [reply]
I will try your suggestions. Thanks.

**Reply by nielsbasjes** [reply]
In my config I also have this one:

    {
      "matchPackagePrefixes": [
        "org.apache.maven.plugins",
        "org.codehaus.mojo",
        "org.sonarsource.scanner.maven",
        "com.github.spotbugs",
        "com.google.code.maven-replacer-plugin",
        "io.fabric8",
        "pl.project13.maven",
        "com.alexecollins.maven.plugin",
        "org.jetbrains.kotlin",
        "net.alchim31.maven"
      ],
      "groupSlug": "maven_build_plugins",
      "groupName": "maven build plugins",
      "labels": [
        "maven-build-plugins"
      ]
    },


and this one

    {
      "description": "Disallow Cloudera/Hortonworks/Spring/... internal versions",
      "matchPackagePrefixes": ["org.apache."],
      "allowedVersions": "/^[0-9]{1,3}\\.[0-9]{1,3}(?:\\.[0-9]{1,3})?(?:-M[0-9]{1,3})?$/"
    },

I suspect double rules that match is indeed the problem I'm facing.
I'm going to change the ordering and see what happens.

**Reply by nielsbasjes** [reply]
@PhilipAbed 
Thank you for your help. It was indeed 2 rules that matched and the ordering of those rules.
After I move the rule after the last one I showed the MR was almost immediately autoclosed.
